File: PHP/6php 3/rank.php
<?php
/*
|--------------------------------------------------------------------------
| Prize Draw
|--------------------------------------------------------------------------
|
*/

function rank($st, $we, $n) {
   if (empty($st)) return "No participants";
   $names = explode(",", $st);
   if ($n > count($names)) return "Not enough participants";
   $participants = array();
   foreach ($names as $i => $name) {
      $som = strlen($name) + array_sum(array_map(function($char) {
         return ord($char) - 64;
      }, str_split(strtoupper($name))));
      $participants[] = array('name' => $name, 'winningNumber' => $som * $we[$i]);
   }
   usort($participants, function($a, $b) {
      if ($b['winningNumber'] == $a['winningNumber']) {
         return strcmp($a['name'], $b['name']);
      }
      return $b['winningNumber'] - $a['winningNumber'];
   });
   return $participants[$n - 1]['name'];
}


function rank($st, $we, $n) {
   $arr = explode(',', $st);
   if (strlen($st) === 0) return 'No participants';
   if (count($arr) < $n) return 'Not enough participants';
   foreach ($arr as $j => $person) {
      $score = strlen($person);
      for ($i = 0; $i < strlen($person); $i++) {
         $score += (ord(strtolower($person[$i])) - 96);
      }
      $scores[$person] = $score * $we[$j];
   }
   ksort($scores);
   arsort($scores);
   $keys = array_keys($scores);
   return $keys[$n - 1];
}


function rank($st, $we, $n) {
   if ($st === '') return 'No participants';
   $names = explode(',', $st);
   if ($n > count($names)) return 'Not enough participants';
   $win = array();
   foreach ($names as $pos => $name) {
      $sum = strlen($name);
      foreach (str_split($name) as $c) $sum += (ord(strtoupper($c)) - ord('A') + 1);
      $sum *= $we[$pos];
      $win[$name] = $sum;
   }
   ksort($win);
   arsort($win);
   $winner = array_slice($win, $n - 1, 1, true);
   return key($winner);
}


print_r(rank("Addison,Jayden,Sofia,Michael,Andrew,Lily,Benjamin", [4, 2, 1, 4, 3, 1, 2], 4) . " Benjamin
")
print_r(rank("Elijah,Chloe,Elizabeth,Matthew,Natalie,Jayden", [1, 3, 5, 5, 3, 6], 2) . " Matthew
")
print_r(rank("Aubrey,Olivai,Abigail,Chloe,Andrew,Elizabeth", [3, 1, 4, 4, 3, 2], 4) . " Abigail
")
print_r(rank("Lagon,Lily", [1, 5], 2) . " Lagon
")

File: Python/6py 2/rank.py
"""
|--------------------------------------------------------------------------
| Prize Draw
|--------------------------------------------------------------------------
|
"""

def rank(st, we, n):
   if not st: return "No participants"
   names = st.split(",")
   if n > len(names): return "Not enough participants"
   participants = []
   for i, name in enumerate(names):
      som = len(name) + sum(ord(char) - 64 for char in name.upper())
      participants.append({'name': name, 'winningNumber': som * we[i]})
   participants.sort(key=lambda x: (-x['winningNumber'], x['name']))
   return participants[n - 1]['name']


def rank(st, we, n):
   if not st: return "No participants"
   if n>len(we): return "Not enough participants"
   name_score = lambda name,w: w*(len(name)+sum([ord(c.lower())-96for c in name]))
   scores= [name_score(s,we[i]) for i,s in enumerate(st.split(','))]
   scores = list(zip(st.split(','),scores))    
   scores.sort(key=lambda x: (-x[1],x[0]))
   return scores[n-1][0]


def rank(st, we, n):
   mas = [{i : len(i) + sum([ord(j.upper()) - 64 for j in i])} for i in st.split(",")]
   wining_number = {j : mas[i][j] * we[i] for i in range(len(mas)) for j in mas[i]}
   try:
      return sorted(wining_number, key = lambda x: (-wining_number[x], x))[n-1]
   except:
      return "Not enough participants" if st else "No participants"


print(rank("Addison,Jayden,Sofia,Michael,Andrew,Lily,Benjamin", [4, 2, 1, 4, 3, 1, 2], 4), "Benjamin")
print(rank("Elijah,Chloe,Elizabeth,Matthew,Natalie,Jayden", [1, 3, 5, 5, 3, 6], 2), "Matthew")
print(rank("Aubrey,Olivai,Abigail,Chloe,Andrew,Elizabeth", [3, 1, 4, 4, 3, 2], 4), "Abigail")
print(rank("Lagon,Lily", [1, 5], 2), "Lagon")